crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,178 +139,266 @@ impl Executor {
+                SubmitStateProj::NotStarted => {
+                    let tx = UnsignedTransaction {
+                        nonce: *this.nonce,
+                        actions: this.bundle.clone(),
+                    }
+                    .into_signed(this.signing_key);
+                    SubmitState::WaitingForSubmit {
+                        fut: submit_tx(this.client.clone(), tx).boxed(),
+                    }
+                }
 
-    /// Wrapper around [`Client::submit_transaction_sync`] with a 1s timeout.
-    async fn submit_transaction_sync(
-        &self,
-        signed_tx: SignedTransaction,
-    ) -> eyre::Result<tx_sync::Response> {
-        tokio::time::timeout(
-            Duration::from_secs(1),
-            self.inner.submit_transaction_sync(signed_tx),
-        )
-        .await
-        .wrap_err("request timed out")?
-        .wrap_err("RPC returned with error")
+                SubmitStateProj::WaitingForSubmit {
+                    fut,
+                } => match ready!(fut.poll(cx)) {
+                    Ok(rsp) => match AbciCode::from_cometbft(rsp.code) {
+                        Some(AbciCode::OK) => {
+                            info!("sequencer responded with AbciCode zero; submission successful");
+                            return Poll::Ready(Ok(*this.nonce + 1));
+                        }
+                        Some(AbciCode::INVALID_NONCE) => {
+                            info!(
+                                "sequencer responded with `invalid nonce` abci code; fetching new \
+                                 nonce"
+                            );
+                            SubmitState::WaitingForNonce {
+                                fut: get_latest_nonce(this.client.clone(), *this.address).boxed(),
+                            }
+                        }
+                        _other => {
+                            warn!(
+                                abci.code = rsp.code.value(),
+                                abci.log = rsp.log,
+                                "sequencer responded with non-zero abci code; the bundle is \
+                                 likely lost",
+                            );
+                            return Poll::Ready(Ok(*this.nonce));
+                        }
+                    },
+                    Err(e) => {
+                        let error: &(dyn std::error::Error + 'static) = e.as_ref();
+                        error!(
+                            error,
+                            "critically failed submitting transaction to sequencer"
+                        );
+                        return Poll::Ready(
+                            Err(e).wrap_err("failed submitting transaction to sequencer"),
+                        );
+                    }
+                },
+
+                SubmitStateProj::WaitingForNonce {
+                    fut,
+                } => match ready!(fut.poll(cx)) {
```

Comment:
if the `waiting_for_nonce` future isn't ready, then the state is unchanged right?

Reply:
Exactly. The macro short circuits, returning `Poll::Pending` if the underlying future is not ready

Reply:
or does `ready!` wait for the future to be ready? the docs on this macro are kinda bad lol

Reply:
wait nvm - this returns `Poll::Pending` lol, i get it now